internal/guest/storage/scsi/scsi.go

```diff
@@ -175,27 +217,44 @@ func Mount(
 			}
 		}
 		source = encryptedSource
-		mountType = "xfs"
 	}
 
-	for {
-		if err := unixMount(source, target, mountType, flags, data); err != nil {
-			// The `source` found by GetDevicePath can take some time
-			// before its actually available under `/dev/sd*`. Retry while we
-			// wait for `source` to show up.
-			if errors.Is(err, unix.ENOENT) || errors.Is(err, unix.ENXIO) {
-				select {
-				case <-ctx.Done():
-					log.G(ctx).Warnf("mount system call failed with %s, context timed out while retrying", err)
+	// support mounting any filesystem by checking the device fs present
+	// and using that as the mountType unless `EnsureFilesystem` and `Filesystem`
+	// are set for a different fs to use.
+	deviceFS, err := _getDeviceFsType(source)
+	if err != nil {
+		return err
+	}
+	mountType := deviceFS
+
+	// if EnsureFilesystem is set, then we need to check if the device has the
+	// correct filesystem configured on it. If it does not, format the device
+	// with the corect filesystem. Right now, we only support formatting xfs
+	// and ext4.
+	if config.EnsureFilesystem {
+		if deviceFS != config.Filesystem {
+			// re-format device to the correct fs
+			switch config.Filesystem {
+			case "ext4":
+				if err := ext4Format(ctx, source); err != nil {
+					return err
+				}
+			case "xfs":
+				if err := xfsFormat(source); err != nil {
 					return err
-				default:
-					time.Sleep(10 * time.Millisecond)
-					continue
 				}
+			default:
+				return fmt.Errorf("unsupported filesystem %s requested for device", config.Filesystem)
 			}
-			return err
 		}
-		break
+		mountType = config.Filesystem
+	}
+
+	// device should already be present under /dev, so we should not get an error
+	// unless the command has actually errored out
+	if err := unixMount(source, target, mountType, flags, data); err != nil {
```

Comment:
Shouldn't we mount using `Filesystem` as type here? It seems wrong to ignore the specified filesystem and use the detected one instead.

Reply:
If `Filesystem` is empty, then I would think we would use the detected type.

Reply:
I think I put it in a comment somewhere, but what I decided to do here was make it so that `Filesystem` is only valid if `EnsureFilesystem` is. 


I didn't want to use an empty string for `Filesystem` to indicate that we should use the detected fs since we'd lose the ability to tell the difference between "use what's detected" and "we want the device to have no fs configured." I'm not immediately aware of a need for the latter scenario, however. What do you think?

Reply:
I think it would be _extremely_ confusing to have a `Filesystem` field on a "mount" operation which doesn't determine what filesystem should be used for the mount.

What does it mean to do a mount with "no fs configured"? If you use the mount tool on Linux and don't specify an fs, it auto-detects which to use, which we already do on our end.

Reply:
Also, we had initially talked about supporting a case in the future where the client tells wants to specify exactly what fs to mount, even if it's not done we explicitly detect.

Reply:
Ah that seems fair. I looked back at the docs for mount and you're right in that we can't mount a device that doesn't have a fs on it. In that case I'll change this and update the comments.

Reply:
Updated the code to use the `Filesystem` as the mount type whether or not `EnsureFilesystem` is set and updated comments.